File: 3DEngine/Classes/Delaunay/Point3d.vb
Imports System.Runtime.InteropServices

Namespace Delaunay
    Public Class Point3d
        Implements IEquatable(Of Point3d)

        Public Const ToRad As Double = Math.PI / 180.0
        Public Const ToDeg As Double = 180.0 / Math.PI

        Public Property X As Double
        Public Property Y As Double
        Public Property Z As Double

        Public Sub New()
        End Sub

        Public Sub New(p As Point3d)
            X = p.X
            Y = p.Y
            Z = p.Z
        End Sub

        Public Sub New(x As Double, y As Double, z As Double)
            Me.X = x
            Me.Y = y
            Me.Z = z
        End Sub

        Public Function Distance(p As Point3d) As Double
            Dim dx As Double = X - p.X
            Dim dy As Double = Y - p.Y
            Dim dz As Double = Z - p.Z
            Return Math.Sqrt(dx * dx + dy * dy + dz * dz)
        End Function

        Public Function Length() As Double
            Return Math.Sqrt(X * X + Y * Y + Z * Z)
        End Function

        Public Function Dot(p As Point3d) As Double
            Return X * p.X + Y * p.Y + Z * p.Z
        End Function

        Public Function Cross(p As Point3d) As Point3d
            Return New Point3d((Y * p.Z) - (p.Y * Z),
                               (Z * p.X) - (p.Z * X),
                               (X * p.Y) - (p.X * Y))
        End Function

        Public Sub Normalize()
            Dim len As Double = Me.Length()
            If len > 0 Then
                X /= len
                Y /= len
                Z /= len
            Else
                X = 1.0
                Y = 0.0
                Z = 0.0
            End If
        End Sub

        Public Function Normalized() As Point3d
            Dim p As New Point3d(Me)
            p.Normalize()
            Return p
        End Function

        Public Function RotateX(a As Double) As Point3d
            Dim arad As Double = a * ToRad
            Dim cosa As Double = Math.Cos(arad)
            Dim sina As Double = Math.Sin(arad)
            Return New Point3d(X, Y * cosa - Z * sina, Y * sina + Z * cosa)
        End Function

        Public Function RotateY(a As Double) As Point3d
            Dim arad As Double = a * ToRad
            Dim cosa As Double = Math.Cos(arad)
            Dim sina As Double = Math.Sin(arad)
            Return New Point3d(Z * sina + X * cosa, Y, Z * cosa - X * sina)
        End Function

        Public Function RotateZ(a As Double) As Point3d
            Dim arad As Double = a * ToRad
            Dim cosa As Double = Math.Cos(arad)
            Dim sina As Double = Math.Sin(arad)
            Return New Point3d(X * cosa - Y * sina, X * sina + Y * cosa, Z)
        End Function

        Public Function Project(viewWidth As Integer, viewheight As Integer, fov As Double, viewDistance As Double) As Point3d
            Dim factor As Double = If(viewDistance = -Z, 999, fov / (viewDistance + Z))
            Return New Point3d(X * factor + viewWidth / 2, Y * factor + viewheight / 2, Z)
        End Function

        Public Function UnProject(viewWidth As Integer, viewheight As Integer, fov As Double, viewDistance As Double) As Point3d
            Dim factor As Double = If(viewDistance = -Z, 999, fov / (viewDistance + Z))
            Return New Point3d((X - viewWidth / 2) / factor, (Y - viewheight / 2) / factor, Z)
        End Function

        Public Function ToPointF() As PointF
            Return New PointF(X, Y)
        End Function

        Public Shared Operator =(p1 As Point3d, p2 As Point3d) As Boolean
            Return p1.IsSimilar(p2)
            'Return p1.X = p2.X AndAlso p1.Y = p2.Y AndAlso p1.Z = p2.Z
        End Operator

        Public Shared Operator <>(p1 As Point3d, p2 As Point3d) As Boolean
            Return Not (p1 = p2)
        End Operator

        Public Shared Operator +(p1 As Point3d, p2 As Point3d) As Point3d
            Return New Point3d(p1.X + p2.X, p1.Y + p2.Y, p1.Z + p2.Z)
        End Operator

        Public Shared Operator -(p1 As Point3d, p2 As Point3d) As Point3d
            Return New Point3d(p1.X - p2.X, p1.Y - p2.Y, p1.Z - p2.Z)
        End Operator

        Public Shared Operator -(p1 As Point3d) As Point3d
            Return New Point3d(p1.X, p1.Y, -p1.Z)
        End Operator

        Public Shared Operator *(p1 As Point3d, scalar As Double) As Point3d
            Return New Point3d(p1.X * scalar, p1.Y * scalar, p1.Z * scalar)
        End Operator

        Public Shared Operator *(scalar As Double, p1 As Point3d) As Point3d
            Return p1 * scalar
        End Operator

        Public Shared Operator *(p1 As Point3d, p2 As Point3d) As Double
            Return p1.Dot(p2)
        End Operator

        Public Shared Operator /(p1 As Point3d, scalar As Double) As Point3d
            Return p1 * (1 / scalar)
        End Operator

        Public Function ToArray() As Double()
            Return {X, Y, Z}
        End Function

        Public Overrides Function ToString() As String
            Return $"({X:F2}, {Y:F2}, {Z:F2})"
        End Function

        Public Function AsInt(Optional padding As Integer = 0) As Point3d
            Dim X1 As Integer = X
            Dim Y1 As Integer = Y
            Dim Z1 As Integer = Z

            If padding > 1 Then
                X1 -= X1 Mod padding
                Y1 -= Y1 Mod padding
                Z1 -= Z1 Mod padding
            End If

            Return New Point3d(X1, Y1, Z1)
        End Function

        Public Function Compare(p As Point3d) As Integer
            Return Length.CompareTo(p.Length)
        End Function

        Public Shadows Function Equals(other As Point3d) As Boolean Implements IEquatable(Of Point3d).Equals
            Return Me = other
        End Function

        Public Function IsSimilar(p As Point3d) As Boolean
            Return Math.Abs(X - p.X) <= Triangualtor.Epsilon AndAlso
                   Math.Abs(Y - p.Y) <= Triangualtor.Epsilon AndAlso
                   Math.Abs(Z - p.Z) <= Triangualtor.Epsilon
        End Function

        Public Function AngleXY(p As Point3d) As Double
            Dim a As Double = Math.Atan2(p.Y - Y, p.X - X) * ToDeg
            If a < 0 Then a += 360.0
            Return a
        End Function

        Public Function AngleXZ(p As Point3d) As Double
            Dim a As Double = Math.Atan2(p.Z - Z, p.X - X) * ToDeg
            If a < 0 Then a += 360.0
            Return a
        End Function

        Public Function AngleYZ(p As Point3d) As Double
            Dim a As Double = Math.Atan2(p.Z - Z, p.Y - Y) * ToDeg
            If a < 0 Then a += 360.0
            Return a
        End Function

        ' Unfortunately, this doesn't seem to improve squat!
        'Private Const PI2 As Double = Math.PI / 2
        'Private Const B As Double = 4 / Math.PI
        'Private Const C As Double = -4 / (Math.PI * Math.PI)

        'Public Shared Function FastSin(a As Double) As Double
        '    Return B * a + C * a * Math.Abs(a)
        'End Function

        'Public Shared Function FastCos(a As Double) As Double
        '    Return FastSin(a + PI2)
        'End Function
    End Class
End Namespace

File: 3DEngine/Delaunay/Point3d.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Delaunay {
    public class Point3d {
        public const double ToRad = Math.PI / 180.0;
        public const double ToDeg = 180.0 / Math.PI;

        public double X;
        public double Y;
        public double Z;

        public Point3d() {
        }

        public Point3d(Point3d p) {
            X = p.X;
            Y = p.Y;
            Z = p.Z;
        }

        public Point3d(double x, double y, double z) {
            X = x;
            Y = y;
            Z = z;
        }

        public double Distance(Point3d v) {
            double dx = X - v.X;
            double dy = Y - v.Y;
            double dz = Z - v.Z;
            return Math.Sqrt(dx * dx + dy * dy + dz * dz);

            //return Math.Sqrt(Math.Pow((this.X - v.X), 2.0) +
            //                 Math.Pow((this.Y - v.Y), 2.0) +
            //                 Math.Pow((this.Z - v.Z), 2.0));
        }

        public double Distance2d(Point3d v) {
            return Math.Sqrt(Math.Pow((this.X - v.X), 2.0) +
                             Math.Pow((this.Y - v.Y), 2.0));
        }

        public double Length() {
            return Math.Sqrt(X * X + Y * Y + Z * Z);
        }

        public double Dot(Point3d v) {
            return X * v.X + Y * v.Y + Z * v.Z;
        }

        public Point3d Cross(Point3d v) {
            return new Point3d((Y * v.Z) - (v.Y * Z),
                               (Z * v.X) - (v.Z * X),
                               (X * v.Y) - (v.X * Y));
        }

        public void Normalize() {
            double len = this.Length();
            if(len > 0) {
                X /= len;
                Y /= len;
                Z /= len;
            } else {
                X = 1.0;
                Y = 0.0;
                Z = 0.0;
            }
        }

        public Point3d Normalized() {
            Point3d p = new Point3d(this);
            p.Normalize();
            return p;
        }

        public Point3d RotateX(double angle) {
            double rad = angle * ToRad;
            double cosa = Math.Cos(rad);
            double sina = Math.Sin(rad);
            double yn = Y * cosa - Z * sina;
            double zn = Y * sina + Z * cosa;

            return new Point3d(X, yn, zn);
        }

        public Point3d RotateY(double angle) {
            double rad = angle * ToRad;
            double cosa = Math.Cos(rad);
            double sina = Math.Sin(rad);
            double xn = Z * sina + X * cosa;
            double zn = Z * cosa - X * sina;

            return new Point3d(xn, Y, zn);
        }

        public Point3d RotateZ(double angle) {
            double rad = angle * ToRad;
            double cosa = Math.Cos(rad);
            double sina = Math.Sin(rad);
            double xn = X * cosa - Y * sina;
            double yn = X * sina + Y * cosa;

            return new Point3d(xn, yn, Z);
        }

        public Point3d Project(int viewWidth, int viewHeight, double fov, double viewDistance) {
            double factor = (viewDistance == -Z ? 999 : fov / (viewDistance + Z));
            return new Point3d(X * factor + viewWidth / 2, Y * factor + viewHeight / 2, Z);
        }

        public Point3d UnProject(int viewWidth, int viewHeight, double fov, double viewDistance) {
            double factor = (viewDistance == -Z ? 999 : fov / (viewDistance + Z));
            return new Point3d((X - viewWidth / 2) / factor, (Y - viewHeight / 2) / factor, Z);
        }

        public System.Drawing.PointF ToPointF() {
            return new System.Drawing.PointF((float)this.X, (float)this.Y);
        }

        public override bool Equals(object obj) {
            if(obj is Point3d) {
                Point3d p = (Point3d)obj;
                return (this.X == p.X && this.Y == p.Y && this.Z == p.Z);
            } else {
                return false;
            }
        }

        public override int GetHashCode() {
            return base.GetHashCode();
        }

        public bool IsSimilar(Point3d v, double epsilon = 0.0001) {
            return Math.Abs(X - v.X) <= epsilon &&
                   Math.Abs(Y - v.Y) <= epsilon &&
                   Math.Abs(Z - v.Z) <= epsilon;
        }

        public static bool operator ==(Point3d p1, Point3d p2) {
            //if(p1 == null && p2 != null) return false;
            //if(p2 == null && p1 != null) return false;
            //if(p1 == null && p2 == null) return true;
            return (p1?.X == p2?.X && p1?.Y == p2?.Y && p1?.Z == p2?.Z);
        }

        public static bool operator !=(Point3d p1, Point3d p2) {
            return !(p1 == p2);
        }

        public static Point3d operator +(Point3d v1, Point3d v2) {
            return new Point3d(v1.X + v2.X, v1.Y + v2.Y, v1.Z + v2.Z);
        }

        public static Point3d operator -(Point3d v1, Point3d v2) {
            return new Point3d(v1.X - v2.X, v1.Y - v2.Y, v1.Z - v2.Z);
        }

        public static Point3d operator -(Point3d p1) {
            return new Point3d(p1.X, p1.Y, -p1.Z);
        }

        public static Point3d operator *(Point3d v1, double scalar) {
            return new Point3d(v1.X * scalar, v1.Y * scalar, v1.Z * scalar);
        }

        public static Point3d operator *(double scalar, Point3d v1) {
            return new Point3d(v1.X * scalar, v1.Y * scalar, v1.Z * scalar);
        }

        public static double operator *(Point3d v1, Point3d v2) {
            return v1.Dot(v2);
        }

        public static Point3d operator /(Point3d v1, double scalar) {
            return v1 * (1 / scalar);
        }

        public override string ToString() {
            return string.Format("({0:F2}, {1:F2}, {2:F2})", X, Y, Z);
        }

        public Point3d AsInt(int padding = 0) {
            int x = (int)Math.Round(this.X);
            int y = (int)Math.Round(this.Y);
            int z = (int)Math.Round(this.Z);

            if(padding > 1) {
                x -= x % padding;
                y -= y % padding;
                z -= z % padding;
            }

            return new Point3d(x, y, z);
        }

        public double AngleXY(Point3d v1) {
            double dx = v1.X - this.X;
            double dy = v1.Y - this.Y;

            double a = Math.Atan2(dy, dx) * ToDeg;
            if(a < 0) a += 360;
            return a;
        }

        public double AngleXZ(Point3d v1) {
            double dx = v1.X - this.X;
            double dz = v1.Z - this.Z;

            double a = Math.Atan2(dz, dx) * ToDeg;
            if(a < 0) a += 360;
            return a;
        }

        public double AngleYZ(Point3d v1) {
            double dy = v1.Y - this.Y;
            double dz = v1.Z - this.Z;

            double a = Math.Atan2(dz, dy) * ToDeg;
            if(a < 0) a += 360;
            return a;
        }
    }
}
